### PR TITLE
Include JDKs in Dists

### PR DESCRIPTION
**Pull request by CRogers**
## Before this PR
We want to include JDKs in dists (reasons in [this doc](https://pl.ntr/2hO)). There is currently no way to do this.

I had tried to do this all from an internal plugin, but it's impossible to set `javaHome` correctly and maintain laziness.

## After this PR
We add a way to include a number of JDKs to a dist, up to one per version, and then set `javaHome` in go-java-launcher to the correct one.

Multiple JDKs are required as some internal services run user code, which select from multiple JDK versions, and these should continue to work. I'm assuming we won't need multiple _variants_ of the same JDK version, as that would be hugely wasteful to deploy and we should avoid this as much as we can.

==COMMIT_MSG==
Add `jdks` property to `distribution` extension to help include JDKs into dists.
==COMMIT_MSG==

### CODE REVIEW

File: gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.java [hunk 1]
```diff
@@ -85,7 +97,13 @@ public JavaServiceDistributionExtension(Project project) {
         excludeFromVar = objectFactory.listProperty(String.class);
         excludeFromVar.addAll("log", "run");
 
-        env = objectFactory.mapProperty(String.class, String.class).empty();
+        env = objectFactory.mapProperty(String.class, String.class);
```
Comment (felixdesouza) [inline gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.java:99]: I think you want to separate out env that's editable by everyone else and this env. I think you're able to actually empty this `env` and everything is fine.

I think you'd want to have a resulting env, that first applies the jdks env and then the user env?

File: gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.java [hunk 2]
```diff
@@ -85,7 +97,13 @@ public JavaServiceDistributionExtension(Project project) {
...
+        env.putAll(project.provider(() -> {
+            return jdks.get().entrySet().stream()
```
Comment (felixdesouza) [inline gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.java]: can't you map `jdks`?

Comment (CRogers) [reply gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.java]: Already have - I thought you'd notice that heh

File: gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.java (hunk 1 above)
Comment (CRogers) [reply gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.java:99]: True. No one uses `env` internally at the moment, but I guess it should still not be editable.

Comment (CRogers) [reply gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.java:99]: fixed + test

File: gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java [hunk 3]
```diff
@@ -331,7 +331,7 @@ private static Provider<Map<String, String>> userConfiguredEnvWithJdkEnvVars(
         return distributionExtension.getEnv().zip(distributionExtension.getJdks(), (userConfiguredEnv, jdks) -> {
             Map<String, String> actualEnv = new HashMap<>(userConfiguredEnv);
```
Comment (felixdesouza) [inline gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java]: wouldn't you need a `LinkedHashMap` in addition to the below

Comment (CRogers) [reply gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java]: I just thought that... very tricksy

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPluginTests.groovy [hunk 4]
```diff
@@ -179,7 +179,6 @@ class JavaServiceDistributionPluginTests extends GradleIntegrationSpec {
             }
 
             repositories {
-                jcenter()
```
Comment (CRogers) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPluginTests.groovy:182]: Deprecation warning with new higher version of gradle

File: gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/SlsBaseDistPlugin.java [hunk 5]
```diff
@@ -31,7 +31,7 @@ public class SlsBaseDistPlugin implements Plugin<Project> {
 
     public static final String SLS_DIST_USAGE = "sls-dist";
 
-    public static final GradleVersion MINIMUM_GRADLE = GradleVersion.version("6.1");
+    public static final GradleVersion MINIMUM_GRADLE = GradleVersion.version("7.6");
```
Comment (CRogers) [inline gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/SlsBaseDistPlugin.java:34]: `Provider#zip` is introduced in 6.6. However, it seems that gradle introduced plugin bytecode rewriting sometime between 6.1 and 6.6. This means 6.6 now falls over on the jackson 2.15.2 dep from baseline, so we need to go all the way to 7.6

Comment (felixdesouza) [reply gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/SlsBaseDistPlugin.java:34]: don't we need 7.6.2?

Comment (CRogers) [reply gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/SlsBaseDistPlugin.java:34]: 7.6 for Java 19: https://docs.gradle.org/current/userguide/compatibility.html

Comment (felixdesouza) [reply gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/SlsBaseDistPlugin.java:34]: https://docs.gradle.org/7.6.2/release-notes.html and https://github.com/gradle/gradle/issues/24390

Comment (felixdesouza) [reply gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/SlsBaseDistPlugin.java:34]: happy to merge if 7.6 doesn't break and have a lower min version

Comment (CRogers) [reply gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/SlsBaseDistPlugin.java:34]: Huh, I wonder why it works...

**Comment by felixdesouza** [comment]
👍

**Pull request merged**